### PR TITLE
Initial addition of kokoro and bazel continuous integration support.

### PR DESCRIPTION
**Pull request by jverkoey**
Only adds support for ActivityIndicator and its dependencies. Will add other components to the continuous integration in follow up changes.

As part of this change we will need to start consistently using more specific import statements throughout our component tests, e.g. `import MaterialComponents.MaterialActivityIndicator`. The reason we need to do this is because bazel's module names differ from CocoaPods'. We do pre&post-processing on the swift tests during continuous integration to transform `import MaterialComponents.MaterialActivityIndicator` into `import components_ActivityIndicator_ActivityIndicator`.

### CODE REVIEW

**Comment by jverkoey** [comment]
Blocked by https://github.com/material-foundation/material-internationalization-ios/pull/30.

**Comment by jverkoey** [comment]
Boom, 17 minute turnaround! 10 minutes was spent cloning the repo.

File: .kokoro
Comment (romoore) [inline .kokoro]: I believe the copyright should be "the Material Components for iOS authors."

File: BUILD
Comment (romoore) [inline BUILD]: Same copyright authorship feedback.

File: WORKSPACE
Comment (romoore) [inline WORKSPACE]: Copyright again.

File: WORKSPACE
Comment (romoore) [inline WORKSPACE]: Should this URL point to a Material organizational account?  It looks like it's currently pointing to your fork.

File: components/ActivityIndicator/BUILD
Comment (romoore) [inline components/ActivityIndicator/BUILD]: Same copyright feedback.

File: components/ActivityIndicator/BUILD
Comment (romoore) [inline components/ActivityIndicator/BUILD]: Does this automatically pick-up `src/private`?

File: components/ActivityIndicator/BUILD
Comment (romoore) [inline components/ActivityIndicator/BUILD]: Maybe also add headers?

```
hdrs = glob([
    "tests/unit/*.h",
]),
```

File: components/AnimationTiming/BUILD
Comment (romoore) [inline components/AnimationTiming/BUILD]: Copyright again.

File: components/AnimationTiming/BUILD
Comment (romoore) [inline components/AnimationTiming/BUILD]: No sources or headers for this library?

File: components/Palettes/BUILD
Comment (romoore) [inline components/Palettes/BUILD]: Copyright again.

File: components/private/Application/BUILD
Comment (romoore) [inline components/private/Application/BUILD]: Copyright authorship here, too.

**Review by romoore — CHANGES_REQUESTED** [review]
Mostly feedback about copyright headers.

File: .kokoro
Comment (jverkoey) [reply .kokoro]: Done.

File: WORKSPACE
Comment (jverkoey) [reply WORKSPACE]: Done.

File: BUILD
Comment (jverkoey) [reply BUILD]: Done.

File: WORKSPACE
Comment (jverkoey) [reply WORKSPACE]: Done.

File: components/ActivityIndicator/BUILD
Comment (jverkoey) [reply components/ActivityIndicator/BUILD]: Done.

File: components/ActivityIndicator/BUILD
Comment (jverkoey) [reply components/ActivityIndicator/BUILD]: `includes` defines the headers that clients using this dependency will have access to, so we intentionally only reveal the `src` directory.

File: components/ActivityIndicator/BUILD
Comment (jverkoey) [reply components/ActivityIndicator/BUILD]: Done.

File: components/AnimationTiming/BUILD
Comment (jverkoey) [reply components/AnimationTiming/BUILD]: Done.

File: components/AnimationTiming/BUILD
Comment (jverkoey) [reply components/AnimationTiming/BUILD]: Done.

File: components/private/Application/BUILD
Comment (jverkoey) [reply components/private/Application/BUILD]: Done.

File: components/Palettes/BUILD
Comment (jverkoey) [reply components/Palettes/BUILD]: Done.

File: components/ActivityIndicator/BUILD
Comment (romoore) [reply components/ActivityIndicator/BUILD]: Ack.

File: components/AnimationTiming/BUILD
Comment (romoore) [inline components/AnimationTiming/BUILD]: I believe the "*.h" glob should be in hdrs, no?

**Review by ajsecord — CHANGES_REQUESTED** [review]
Hey, so clearly we're duplicating a lot of the internal build configuration here. Is there no way to make this more consistent? E.g. naming of our public objc library Skylark rules?

File: .kokoro
Comment (randallli) [inline .kokoro:32]: does running kokoro result in file system changes?

File: components/ActivityIndicator/BUILD
Comment (randallli) [inline components/ActivityIndicator/BUILD:69]: lets add a build test

File: components/ActivityIndicator/BUILD
Comment (randallli) [inline components/ActivityIndicator/BUILD]: why did you choose to mark each private instead of a global private default?
package(default_visibility = ["//visibility:private"])

**Comment by jverkoey** [comment]
Added a bzl file that matches our internal bazel rule names. We'll separately need to evaluate how to merge the internal rules with the external rules.

File: components/AnimationTiming/BUILD
Comment (jverkoey) [reply components/AnimationTiming/BUILD]: Done.

File: components/ActivityIndicator/BUILD
Comment (randallli) [inline components/ActivityIndicator/BUILD]: we need to add the bundle: MaterialActivityIndicator.bundle
I think you need to make a filegroup and objc_bundle

filegroup(
    name = "BundleFiles",
    srcs = glob([
        "src/MaterialActivityIndicator.bundle/**",
    ]),
)

objc_bundle(
    name = "Bundle",
    bundle_imports = [":BundleFiles"],
)

File: material_components_ios.bzl
Comment (romoore) [inline material_components_ios.bzl]: Nit: "Baze" is misspelled.

File: material_components_ios.bzl
Comment (romoore) [inline material_components_ios.bzl]: Missing warning flags?

File: components/ActivityIndicator/BUILD
Comment (randallli) [inline components/ActivityIndicator/BUILD]: we should also add the sdk_frameworks
    sdk_frameworks = [
        "UIKit",
        "QuartzCore",
    ],

File: material_components_ios.bzl
Comment (romoore) [inline material_components_ios.bzl:44]: Should the `src/private/*.h` glob move down to `hdrs`?

File: material_components_ios.bzl
Comment (romoore) [inline material_components_ios.bzl]: What does this mean, "relative to src"? If I have a file at "src/private/foo.m" and it imports "src/private/foo.h", then it would contain `#import "private/foo.h"` ?

File: components/ActivityIndicator/tests/unit/ActivityIndicatorTests.swift
Comment (randallli) [inline components/ActivityIndicator/tests/unit/ActivityIndicatorTests.swift:18]: I agree this should be discouraged in our clients and banned from our examples.

**Review by randallli — CHANGES_REQUESTED** [review]
I focused on ActivityIndicator build file. We should apply similar changes to the other BUILD files.

File: components/AnimationTiming/BUILD
Comment (romoore) [inline components/AnimationTiming/BUILD]: Move "*.h" in to `hdrs`?

File: components/Palettes/BUILD
Comment (romoore) [inline components/Palettes/BUILD]: Same question/suggestion about moving "*.h" into `hdrs`

File: .kokoro
Comment (jverkoey) [reply .kokoro:32]: While this script is running it needs to perform some amount of source modification in order for swift import statements to point to the bazel naming convention of `path_to_target_target`. The script makes the modifications before running bazel and then reverts the modifications once the script terminates.

File: components/ActivityIndicator/BUILD
Comment (jverkoey) [reply components/ActivityIndicator/BUILD]: Done.

Comment (jverkoey) [reply components/ActivityIndicator/BUILD]: Done.

File: components/ActivityIndicator/BUILD
Comment (jverkoey) [reply components/ActivityIndicator/BUILD:69]: Build tests aren't available in bazel unfortunately.

File: components/AnimationTiming/BUILD
Comment (jverkoey) [reply components/AnimationTiming/BUILD]: Done.

File: components/Palettes/BUILD
Comment (jverkoey) [reply components/Palettes/BUILD]: Done.

File: material_components_ios.bzl
Comment (jverkoey) [reply material_components_ios.bzl]: Done.

File: material_components_ios.bzl
Comment (jverkoey) [reply material_components_ios.bzl]: Done.

File: material_components_ios.bzl
Comment (jverkoey) [reply material_components_ios.bzl]: Yep.

File: material_components_ios.bzl
Comment (jverkoey) [reply material_components_ios.bzl:44]: Nope - `hdrs` defines the files that are allowed to be imported by a client and we don't want clients pulling in private headers.

> `hdrs`: A list of header files. These are the header files that are exported by this rule.

https://docs.bazel.build/versions/master/be/c-cpp.html#cc_inc_library.hdrs

File: material_components_ios.bzl
Comment (romoore) [reply material_components_ios.bzl]: Ack. That seems really awkward. Aren't includes usually either relative to the current file or just matched against files on the include path?

File: material_components_ios.bzl
Comment (romoore) [reply material_components_ios.bzl:44]: Ack. I think this is the link, though:

> The list of C, C++, Objective-C, and Objective-C++ files that are included as headers by source files in this rule or by users of this library. These will be compiled separately from the source if modules are enabled.

https://docs.bazel.build/versions/master/be/objective-c.html#objc_library.hdrs

**Review by romoore — APPROVED** [review]

File: material_components_ios.bzl
Comment (jverkoey) [reply material_components_ios.bzl]: Ah I misunderstood your original question - relative imports do work as expected and I agree that in that light this comment is somewhat misleading.

**Comment by jverkoey** [comment]
@randallli @ajsecord need your lgtm to merge. Note that this PR intentionally doesn't include every single component yet so that we can slowly ease in to this.

**Review by randallli — CHANGES_REQUESTED** [review]
can we confirm the internal import isn't adversely affected by adding these build files?
https://github.com/material-components/material-components-ios/blob/develop/contributing/releasing.md#testing-with-release-blocking-clients

**Comment by jverkoey** [comment]
@randallli we're already set up not to clobber any BUILD files when we import locally. I just sent you cl/174188137 to also ensure that we don't clobber our internal .bzl file.

Eventually we'll want to translate the public BUILD files into the internal BUILD files during import, but I figure we can evaluate that in a follow-up.

File: material_components_ios.bzl
Comment (ajsecord) [reply material_components_ios.bzl]: The contents of these macros aren't Jeff's doing, I wrote them a while ago. Let's take that discussion offline so we keep the internal/external definitions in sync and Jeff doesn't have to guess at what I was trying to do. :)

**Review by ajsecord — APPROVED** [review]
Let's keep iterating on this. I'd like a quick write-up of the plan here, since this seems more involved than our Travis CI work.

**Review by randallli — APPROVED** [review]

**Pull request merged**
